source/nsdynmemLIB/nsdynmemLIB.c

```diff
@@ -290,178 +238,49 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
-                            //debug("USE Same\r\n");
-                            //Move pointer to Allocated data sector
-                            *ptr++ = s_size;
-                            //Set return Pointer
-                            retval = (void *) ptr;
-                            ptr += s_size;//Move Pointer  to next free place
-                            *ptr = s_size;
-                        }
-                        break;
-                    } else {
-                        //Set Number from negative to Positive
-                        s_size = -s_size;
-                    }
-
-                }
-                //Allocated sector
-                //Move
-                //Set Move field
-                moved += (s_size + 2);
-                ptr += 2; //Skip Both of length fileds
-                ptr += s_size; //Move Pointer  to next free place
-                if (heap_main_end == ptr) {
-                    break;
-                }
-            } else {
-                heap_failure(NS_DYN_MEM_HEAP_SECTOR_CORRUPTED);
-                retval = 0;
-                break;
-            }
-        }
-        if (mem_stat_info_ptr) {
-            alloc_size = (alloc_sector * sizeof(int));
-            if (retval) {
-
-                //Update Allocate OK
-                dev_stat_update(DEV_HEAP_ALLOC_OK, alloc_size);
-
-            } else {
-                //Update Allocate Fail
-                dev_stat_update(DEV_HEAP_ALLOC_FAIL, alloc_size);
-            }
-        }
-        platform_exit_critical();
-    }
-    return retval;
-
-#else
-
-    void *retval = 0;
-    if (alloc_size) {
-        platform_enter_critical();
-        retval = malloc(alloc_size);
-        platform_exit_critical();
-    }
-    return retval;
-#endif
+    return ns_dyn_mem_internal_alloc(alloc_size, 1);
 }
 
 #ifndef STANDARD_MALLOC
-static void ns_free_and_merge_next_sector(int *cur_sector, int size)
+static void ns_free_and_merge_with_adjacent_blocks(int *cur_block, int size)
 {
-    int *end_sector;
-    int next_len;
-    end_sector = cur_sector;
-    end_sector += size + 1;
-    if (end_sector != heap_main_end) {
-        end_sector++;
-        next_len = *end_sector;
-
-        if (next_len > 0) {
-            //Convert End of len
-            end_sector--;
-        } else {
-            //Merge First to second
-            //Set First End
-            //Skip Data And sector start of len
-            //debug("Free Next\r\n");
-            next_len = -(next_len);
-            end_sector += (next_len + 1);
-            size += (next_len + 2);
+    // Theory of operation: Block is always in form | Len | Data | Len |
+    // So we need to check length of previous (if current not heap start)
+    // and next (if current not heap end) blocks. Negative length means
+    // free memory so we can merge freed block with those.
+
+    int *start = cur_block;
+    int *end = cur_block + size + 1;
+    int total_size = size;
```

Comment:
Same argument against "total" as above. Maybe "data_size" and "merged_data_size"?